src/lib/reasoners/bitv.ml

```diff
@@ -452,6 +526,38 @@ module Shostak(X : ALIEN) = struct
       let y_v = { sz = s' ; bv = Other id' } in
       [ (x_v , a2 @ c @ a1) ; (y_v , a2' @ c @ a1') ]
 
+    (* Orient the equality [xt[siz]^{i1, i1 + tai} = xt[siz]^{i2, i2 + tai}]
+
+       The [overl] variable contains the number of overlapping bits.
+
+       - If there are no overlapping bits, the two parts are independent.
+          We create a fresh B-variable [b] and A-variables [a1], [a2] and [a3]
+          and orient:
+
+            [xt = a3 @ b @ a2 @ b @ a1]
+
+          The B-variable is used for the common part that is repeated.
+
+       - If there are overlapping bits, we create only the A variables [a1] and
+          [a3] for the before/after parts, and we create two B variables.
+
+          [b_box] is the total constraint size (from [i1] to [i2 + tai]).
+          [nn_overl] is the number of *initial* non-overlapping bits (from [i1]
+          to [i2]).
+
+          The [b] vector is then split into two parts to properly align the
+          repetition. I won't go into the details of the modulo arithmetic, but
+          here is an example ([u] and [vv] are two [1] and [2] size B-variables)
+
+             _ nn_overl = 3
+            / \
+            xxxxxxx???     uvvuvvuvvu???
+            ???yyyyyyy  => ???uvvuvvuvvu
+            \________/
+              b_box = 10
```

Comment:
My understanding is the following:
```
  sz_b1 = b_box % nn_overl (size of b1)
  sz_b2 = nn_overl - sz_b1 (size of b2)
  a1[i] @ ((b1 @ b2) * (b_box/nn_overl)) @ b1 @ a2[siz - tai - i2]
```
So in the example you gave, I would expect the solution to be something like:
```
  xxxxxxx???
  ???yyyyyyy
  uvvuvvuvvu
```
I think there was extra iteration of the while loop in the result of the example you gave.

Reply:
or to be clearer: 
```
  $$$xxxxxxx???###
  $$$???yyyyyyy###
  \\\uvvuvvuvvu///
where a1 = \\\, b1 = u, b2 = vv, a2 = ///
```

Reply:
Yeah, I messed up the number of uvv — I think your way of writing it is clearer, and I will also include your formula. Thanks!